zebra/zebra_mlag.c

```diff
@@ -23,16 +23,559 @@
+ */
+void zebra_mlag_process_mlag_data(uint8_t *data, uint32_t len)
+{
+	struct stream *s = NULL;
+	struct stream *s1 = NULL;
+	int msg_type = 0;
+
+	s = stream_new(ZEBRA_MAX_PACKET_SIZ);
+	if (s)
+		msg_type = zebra_mlag_protobuf_decode_message(&s, data, len);
+
+	if (msg_type <= 0) {
+		/* Something went wrong in decoding */
+		stream_free(s);
+		zlog_err("%s: failed to process mlag data-%d, %u", __func__,
+			 msg_type, len);
+		return;
+	}
+
+	/*
+	 * additional four bytes are for mesasge type
+	 */
+	s1 = stream_new(stream_get_endp(s) + ZEBRA_MLAG_METADATA_LEN);
+	stream_putl(s1, msg_type);
+	stream_put(s1, s->data, stream_get_endp(s));
+	thread_add_event(zrouter.master, zebra_mlag_post_data_from_main_thread,
+			 s1, 0, NULL);
+	stream_free(s);
+}
+
+/**********************End of MLAG Interaction********************************/
+
+/************************MLAG Thread Processing*******************************/
+
+/*
+ * after posting every 1000 packets, MLAG Thread wll be yielded to give CPU
+ * for other threads
+ */
+#define ZEBRA_MLAG_POST_LIMIT 100
+
+/*
+ * Thsi thread reads the clients data from the Gloabl queue and encodes with
+ * protobuf and pass on to the MLAG socket.
+ */
+static int zebra_mlag_thread_handler(struct thread *event)
+{
+	struct stream *s;
+	uint32_t wr_count = 0;
+	uint32_t msg_type = 0;
+	int len = 0;
+
+	wr_count = stream_fifo_count_safe(zrouter.mlag_info.mlag_fifo);
+	if (IS_ZEBRA_DEBUG_MLAG)
+		zlog_debug(":%s: Processing MLAG write, %d messages in queue",
+			   __func__, wr_count);
+
+	zrouter.mlag_info.t_write = NULL;
+	for (wr_count = 0; wr_count < ZEBRA_MLAG_POST_LIMIT; wr_count++) {
+		/* FIFO is empty,wait for teh message to be add */
+		if (stream_fifo_count_safe(zrouter.mlag_info.mlag_fifo) == 0)
+			break;
+
+		s = stream_fifo_pop_safe(zrouter.mlag_info.mlag_fifo);
+		if (!s) {
+			zlog_debug(":%s: Got a NULL Messages, some thing wrong",
+				   __func__);
+			break;
+		}
+
+		zebra_mlag_reset_write_buffer();
+		/*
+		 * Encode the data now
+		 */
+		len = zebra_mlag_protobuf_encode_client_data(s, &msg_type);
+
+		/*
+		 * write to MCLAGD
+		 */
+		if (len > 0)
+			zebra_mlag_private_write_data(mlag_wr_buffer, len);
+
+		/*
+		 * If mesasge type is De-register, send a signal to main thread,
+		 * sothat necessary cleanup will be done by main thread.
+		 */
+		if (msg_type == MLAG_DEREGISTER) {
+			thread_add_event(zrouter.master,
+					 zebra_mlag_terminate_pthread, NULL, 0,
+					 NULL);
+		}
+
+		stream_free(s);
+	}
+
+	if (IS_ZEBRA_DEBUG_MLAG)
+		zlog_debug(":%s: Posted  %d messages to MLAGD", __func__,
+			   wr_count);
+	/*
+	 * Currently there is only message write task is enqueued to this
+	 * thread, yielding was added for future purpose, sothat this thread can
+	 * server other tasks also and in case FIFO is empty, this task will be
+	 * schedule when main thread adds some messages
+	 */
+	if (wr_count >= ZEBRA_MLAG_POST_LIMIT)
```

Comment:
I think you want to check if `stream_fifo_count_safe(zrouter.mlag_info.mlag_fifo)` is nonzero instead? Aside from the fact that it's impossible for `wr_count > ZEBRA_MLAG_POST_LIMIT`, what if there were exactly that many messages on `mlag_fifo`? Then we are needlessly rescheduling ourselves.

Reply:
My intention is to check, is the queue still had elements to process. if yes schedule the thread, otherwise next enqueue will schedule it. yes checking for equality suffice here.

and in case queue had exact "POST_LIMIT" elements, it will be scheduled only once, next time wr_count will be reset to "0"